Custom Drug Entry Sheet UI Improvements

janhavisinghh opened the pull request:
https://app.shortcut.com/simpledotorg/story/5046/custom-drug-entry-sheet-ui-improvements


This PR:
- Removes the focus from dosage when user clicks on frequency
- Saves `null` in the repository if dosage edit text value is equal to "mg"
- Changes from numeric keyboard to the default keyboard for drug dosage edit text
- Uses `joinToString` to correctly display dosage and frequency values in `EditMedicineScreen`

Next PR:
- Make the cursor appear right before 'mg' whenever the sheet is opened

msasikanth commented on app/src/main/java/org/simple/clinic/drugs/selection/DrugListItem.kt:
```diff
@@ -45,9 +45,8 @@ data class ProtocolDrugListItem(
-    val dosage = if(prescribedDrug?.dosage == null) "" else prescribedDrug.dosage
-    val dosageAndFrequency = "$dosage, $medicineFrequency"
+    val medicineFrequency = if (prescribedDrug?.frequency == null) null else medicineFrequencyToLabelMap[prescribedDrug.frequency]!!.label
+    val dosageAndFrequency = listOfNotNull(prescribedDrug?.dosage, medicineFrequency).joinToString()
```
Honestly, for a couple of items, it's just fine to hardcode. Was IDE showing some lint warning?

msasikanth replied:
Here we are ending up creating a new list and then getting string from that. Instead, we can directly just create a string. wdyt?

janhavisinghh replied:
So the benefit of using `joinToString` with `listOfNull` is that it will not show nulls and remove the commas if there is a null string. If we hardcode it we had to handle all the null cases and add empty strings if there was null and also add/remove commas depending on if it had a value or not. So `joinToString` seemed like a better and a more cleaner approach to handle this.

msasikanth replied:
Hmm, that makes sense. But readability wise I don't think this is good, for example, we can create a function that creates the dosage and frequency string and do the string manipulations there.

I feel like this would be more understandable than the current implementation, WDYT? Also, this way we don't have to create a new list of objects and then covert them to a string.

```kt
private fun dosageAndFrequency(dosage: String?, frequency: MedicineFrequency): String {
  return if (dosage != null) {
    "$dosage, $frequency"
  } else {
    "$frequency"
  }
}
```

janhavisinghh replied:
We will have to add a check for frequency as well `val medicineFrequency = if (prescribedDrug?.frequency == null) null else medicineFrequencyToLabelMap[prescribedDrug.frequency]!!.label` because we don't want to display `None` frequencies. 
It just seems like we are doing a lot of things here to do something that those methods handle already. 😅

msasikanth replied:
Cool, sounds good 👍🏾

msasikanth commented on app/src/main/java/org/simple/clinic/drugs/selection/DrugListItem.kt:
```diff
@@ -76,9 +75,8 @@ data class CustomPrescribedDrugListItem(
-    val dosage = prescribedDrug.dosage
-    val dosageAndFrequency = "$dosage, $medicineFrequency"
+    val medicineFrequency = if (prescribedDrug.frequency == null) null else medicineFrequencyToLabelMap[prescribedDrug.frequency]!!.label
+    val dosageAndFrequency = listOfNotNull(prescribedDrug.dosage, medicineFrequency).joinToString()
```
Same comment as above.

msasikanth commented on app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUpdate.kt:
```diff
@@ -76,10 +76,12 @@ class CustomDrugEntryUpdate : Update<CustomDrugEntryModel, CustomDrugEntryEvent,
       patientUuid: UUID,
       dosagePlaceHolder: String
   ): Next<CustomDrugEntryModel, CustomDrugEntryEffect> {
+    val dosage = if (model.dosage == dosagePlaceHolder) null else model.dosage
```
Instead of passing the placeholder, can't we just check if the dosage contains any numeric value or not? WDYT?

janhavisinghh replied:
uumm yeah 👍🏻

janhavisinghh replied:
@msasikanth 
Does using regex to handle this sound like a good approach?

  ```
  val atleastOneDigitRegex = "\\d".toRegex()
    val dosage = if (atleastOneDigitRegex.containsMatchIn(model.dosage!!)) model.dosage else null

```

msasikanth replied:
Yeah, that works. But we can probably go with this approach instead?

```kt
val isAValidDosage = dosage.filter { it.isDigit() }.isNotBlank()
```

msasikanth reviewed (changes requested):
LGTM, left couple of comments PTAL

janhavisinghh commented:
@msasikanth I removed the commits that were adding `dosagePlaceHolder` to `AddMedicineClicked` event and implemented the suggested changes as new commits.

msasikanth reviewed (approved):
LGTM
